src/test/osd/RadosModel.h

```diff
@@ -1059,15 +1068,48 @@ class ReadOp : public TestOp {
     }
     op.getxattrs(&xattrs, 0);
     assert(!context->io_ctx.aio_operate(context->prefix+oid, completion, &op, 0));
+    waiting_on++;
+ 
+    // send 2 pipelined reads on the same object/snap. This can help testing
+    // OSD's read behavior in some scenarios
+    for (uint32_t i = 0; i < 2; ++i) {
+      librados::ObjectReadOperation pipeline_op;
+
+      pipeline_op.read(0,
+                       !old_value.has_contents() ? 0 :
+                       old_value.most_recent_gen()->get_length(old_value.most_recent()),
+                       &pipeline_results[i],
+                       &pipeline_retvals[i]);
+      pipeline_comps[i] = context->rados.aio_create_completion((void *) this, &read_callback, 0);
+      assert(!context->io_ctx.aio_operate(context->prefix+oid, pipeline_comps[i], &pipeline_op, 0));
+      waiting_on++;
+    }
+
     if (snap >= 0) {
       context->io_ctx.snap_set_read(0);
     }
+    context->state_lock.Unlock();
   }
 
   void _finish(CallbackInfo *info)
   {
-    context->state_lock.Lock();
+    Mutex::Locker l(context->state_lock);
     assert(!done);
+    assert(waiting_on > 0);
+    if (--waiting_on) {
+      return;
+    }
+
+    for (vector<librados::AioCompletion *>::iterator i = pipeline_comps.begin();
+         i != pipeline_comps.end(); ++i) {
+      assert((*i)->is_complete());
+      if (int err = (*i)->get_return_value()) {
```

Comment:
Good catch!  I'll make the changes later. For the last 2 suggestion, I'm doing a little different in the original read and the later pipelined reads (not reading the omap and xattrs). If we want to consolidate the checks for all the reads, we'll need to have multiple copies of `attrs`, `attrretval`, `omap_requested_keys`, etc. What do you think?